### PR TITLE
Fix high cpu load

### PR DESCRIPTION
**Pull request by tonynajjar**
image

image

### CODE REVIEW

File: costmap_converter/test/costmap_polygons.cpp [hunk 1]
```diff
@@ -122,10 +122,10 @@ TEST_F(CostmapToPolygonsDBSMCCHTest, dbScan)
   std::vector< std::vector<costmap_converter::CostmapToPolygonsDBSMCCH::KeyPoint> > clusters;
```
Comment (tonynajjar) [inline costmap_converter/test/costmap_polygons.cpp:122]: fixes this build warning

```
/opt/ros/galactic/src/gtest_vendor/include/gtest/gtest.h:1527:11: warning: comparison of integer expressions of different signedness: ‘const int’ and ‘const long unsigned int’ [-Wsign-compare]
 1527 |   if (lhs == rhs) {

```

File: costmap_converter/src/costmap_to_polygons.cpp [hunk 2]
```diff
@@ -215,10 +215,12 @@ void CostmapToPolygonsDBSMCCH::updateCostmap2D()
...
         n.clear();
 
+      auto size_x = costmap_->getSizeInCellsX();
+      auto size_y = costmap_->getSizeInCellsY();
       // get indices of obstacle cells
-      for(std::size_t i = 0; i < costmap_->getSizeInCellsX(); i++)
+      for(std::size_t i = 0; i < size_x; i++)
       {
-        for(std::size_t j = 0; j < costmap_->getSizeInCellsY(); j++)
+        for(std::size_t j = 0; j < size_y; j++)
```
Comment (tonynajjar) [inline costmap_converter/src/costmap_to_polygons.cpp:223]: https://stackoverflow.com/questions/1242185/loop-condition-evaluation

I confirmed with profilling that it is more performant

File: costmap_converter/src/costmap_converter_node.cpp [hunk 3]
```diff
@@ -119,10 +119,9 @@ class CostmapStandaloneConversion : public rclcpp::Node {
 
     if (converter_) {
       converter_->setOdomTopic(odom_topic);
-      converter_->initialize(
-          std::make_shared<rclcpp::Node>("intra_node", "costmap_converter"));
+      converter_->initialize(shared_from_this());
       converter_->startWorker(std::make_shared<rclcpp::Rate>(5),
-                              costmap_ros_->getCostmap(), true);
+                              costmap_ros_->getCostmap(), false);
```
Comment (tonynajjar) [inline costmap_converter/src/costmap_converter_node.cpp]: That's the main reason for high CPU usage. This makes [this function](https://github.com/logivations/costmap_converter/blob/fix-high-CPU/costmap_converter/include/costmap_converter/costmap_converter_interface.h#L243) run which for some reason is really intense.

File: costmap_converter/src/costmap_converter_node.cpp [hunk 4]
```diff
@@ -119,10 +119,9 @@ class CostmapStandaloneConversion : public rclcpp::Node {
 
     if (converter_) {
       converter_->setOdomTopic(odom_topic);
-      converter_->initialize(
-          std::make_shared<rclcpp::Node>("intra_node", "costmap_converter"));
+      converter_->initialize(shared_from_this());
```
Comment (tonynajjar) [inline costmap_converter/src/costmap_converter_node.cpp]: we don't need to do this but I don't see the point in creating another node

File: costmap_converter/src/costmap_converter_node.cpp (hunk 3 above)
Comment (jplapp) [reply costmap_converter/src/costmap_converter_node.cpp]: possibly this while loop just spins too much, adding a Rate might also fix it.

If everything works without spinning at all, then we can also simply not do it. I would be worried that e.g. some subscribers stop working then

File: costmap_converter/src/costmap_to_polygons.cpp (hunk 2 above)
Comment (jplapp) [reply costmap_converter/src/costmap_to_polygons.cpp:223]: nice

File: costmap_converter/src/costmap_converter_node.cpp (hunk 4 above)
Comment (jplapp) [reply costmap_converter/src/costmap_converter_node.cpp]: yes, also DDS is happy if there are fewer nodes

**Review by jplapp — APPROVED** [review]
awesome!

**Comment by tonynajjar** [comment]
I'd like to propose this improvement to upstream, it's quite an important performance improver

**Comment by jplapp** [comment]
yes, definitely. I suspect that this is a subtle change from ros1 to ros2

**Comment by tonynajjar** [comment]
Mmm I can't really propose the fix as is, because it's not a fix, it's just not using a "feature". The feature of spinning in a thread is still broken. 

Fixing the feature would take more work:

Option 1: add a rate like you proposed.

Option 2: use spin and refactor what is intended to be done with need_to_terminate_ using Context shutdown

I would prefer Option 2 because it doesn't seem right to artificially throttle spin_some (it should be done internally)

**Comment by tonynajjar** [comment]
> it doesn't seem right to artificially throttle spin_some (it should be done internally)

actually I'm seeing this done elsewhere, seems legit

**Comment by tonynajjar** [comment]
https://github.com/rst-tu-dortmund/costmap_converter/pull/36

**Pull request merged**